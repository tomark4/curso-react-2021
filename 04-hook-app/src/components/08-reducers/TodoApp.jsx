import { useEffect, useReducer } from 'react';
import todoReducer from './todoReducer';
import './reducers.css';
import TodoItem from './TodoItem';
import TodoAdd from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

const TodoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    },[todos])

    const handleDelete = (todoId) => {
        dispatch({
            type: 'DELETE',
            payload: todoId
        });
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'TOGGLE',
            payload: todoId
        });
    }

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'ADD',
            payload: newTodo
        });
    }

    return (
        <div className="container my-5">
            <h3>TODO app</h3>
            <hr/>
            <p className="my-2">todos en mi lista: { todos.length }</p>
            <div className="row">
                <div className="col-md-8">
                    {todos.length === 0 &&
                        <div className="alert alert-info">No hay todos</div>
                    }
                    <ul className="list-group list-group-flush">
                        {todos.map( todo => (
                            <TodoItem todo={todo}
                            key={todo.id}
                            onDelete={handleDelete}
                            onTouch={handleToggle}
                            />
                        ))}
                    </ul>
                </div>
                <div className="col-md-4">
                    <TodoAdd onAdd={handleAddTodo} />
                </div>
            </div>
        </div>
    )
}

export default TodoApp
