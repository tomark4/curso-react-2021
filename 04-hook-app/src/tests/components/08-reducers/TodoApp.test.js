import TodoApp from '../../../components/08-reducers/TodoApp';
import {shallow, mount} from 'enzyme';
import {demoTodos} from '../../fixtures/demoTodo';
import { act } from '@testing-library/react';

describe('pruebas en todoapp', () => {

  const wrapper = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn()

  test('should de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('should de agregar un TODO', () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find('TodoAdd').prop("onAdd")(demoTodos[0]);
      wrapper.find('TodoAdd').prop("onAdd")(demoTodos[1]);
    })

    expect( wrapper.find("p.my-2").text().trim() ).toBe("todos en mi lista: 2");
    expect( localStorage.setItem ).toHaveBeenCalledTimes(2)
  });

  test('should debe de eliminar un todo', () => {
    wrapper.find('TodoAdd').prop("onAdd")(demoTodos[0]);
    expect( wrapper.find("p.my-2").text().trim() ).toBe("todos en mi lista: 1");

    wrapper.find("TodoItem").prop('onDelete')(1)
    expect( wrapper.find("p.my-2").text().trim() ).toBe("todos en mi lista: 0");
  })

})

