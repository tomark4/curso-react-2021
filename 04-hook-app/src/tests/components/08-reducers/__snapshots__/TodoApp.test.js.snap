// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en todoapp should de mostrarse correctamente 1`] = `
<div
  className="container my-5"
>
  <h3>
    TODO app
  </h3>
  <hr />
  <p
    className="my-2"
  >
    todos en mi lista: 
    0
  </p>
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="alert alert-info"
      >
        No hay todos
      </div>
      <ul
        className="list-group list-group-flush"
      />
    </div>
    <div
      className="col-md-4"
    >
      <TodoAdd
        onAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
